README.md
# CurrencyFormatter
![](https://img.shields.io/hexpm/v/currency_formatter.svg) ![](https://img.shields.io/hexpm/dt/currency_formatter.svg) ![](https://img.shields.io/hexpm/dw/currency_formatter.svg) ![](https://img.shields.io/coveralls/smeevil/currency_formatter.svg) ![](https://img.shields.io/github/issues/smeevil/currency_formatter.svg) ![](https://img.shields.io/github/issues-pr/smeevil/currency_formatter.svg) ![](https://semaphoreci.com/api/v1/smeevil/currency_formatter/branches/master/shields_badge.svg)

This package offers an Elixir function to format a number to a currency using ISO standards.

The JSON iso data has been gracefully borrowed from the [ruby money](https://github.com/RubyMoney/money/blob/master/config/currency_iso.json) gem.

## Examples :

Formatting cents to currency string
```elixir
iex> CurrencyFormatter.format(654321, :eur)
"€6.543,21"

iex> CurrencyFormatter.format(654300, :eur)
"€6.543"

iex> CurrencyFormatter.format(654300, :eur, keep_decimals: true)
"€6.543,00"


iex> CurrencyFormatter.format(123456)
"$1,234.56"

iex> CurrencyFormatter.format(654321, "AUD")
"$6,543.21"

iex> CurrencyFormatter.format(123456, disambiguate: true)
"US$1,234.56"

iex> CurrencyFormatter.format(654321, "AUD", disambiguate: true)
"A$6,543.21"
```

Formatting cents to a currency raw html string
```elixir
      iex> CurrencyFormatter.raw_html_format(123456, "EUR")
      ~s[<span class="currency-formatter-symbol">€</span><span class="currency-formatter-amount">1.234,56</span>]
  """
```

Formatting cents to a currency safe(phoenix) html string

```elixir
iex> CurrencyFormatter.html_format(123456, "EUR")
      [
        safe: [
          60,
          "span",
          [[32, "class", 61, 34, "currency-formatter-symbol", 34]],
          62,
          "€",
          60,
          47,
          "span",
          62
        ],
        safe: [
          60,
          "span",
          [[32, "class", 61, 34, "currency-formatter-amount", 34]],
          62,
          "1.234,56",
          60,
          47,
          "span",
          62
        ]
      ]
```

Requesting formatting instructions for a currency

```elixir
iex> CurrencyFormatter.instructions(:EUR)
%{"alternate_symbols" => [], "decimal_mark" => ",", "html_entity" => "&#x20AC;",
  "iso_code" => "EUR", "iso_numeric" => "978", "name" => "Euro", "priority" => 2,
  "smallest_denomination" => 1, "subunit" => "Cent", "subunit_to_unit" => 100,
  "symbol" => "€", "symbol_first" => true, "thousands_separator" => "."}
```

Get a map of all currencies and their instructions
```elixir
iex> currencies = CurrencyFormatter.get_currencies()
iex> Enum.count(currencies)
172
iex> currencies["usd"]
%{"alternate_symbols" => ["US$"], "decimal_mark" => ".",
  "disambiguate_symbol" => "US$", "html_entity" => "$", "iso_code" => "USD",
  "iso_numeric" => "840", "name" => "United States Dollar", "priority" => 1,
  "smallest_denomination" => 1, "subunit" => "Cent", "subunit_to_unit" => 100,
  "symbol" => "$", "symbol_first" => true, "thousands_separator" => ","}
```

Getting a list of tuples for use with a select dropdown
```elixir
iex> CurrencyFormatter.get_currencies_for_select()
["AED", "AFN", "ALL", ...]
```

```elixir
iex> CurrencyFormatter.get_currencies_for_select(:names)
[{"AED", "United Arab Emirates Dirham"}, {"AFN", "Afghan Afghani"} , {"ALL", "Albanian Lek"}, ...]

```
```elixir
iex> CurrencyFormatter.get_currencies_for_select(:symbols)
[{"AED", "د.إ"}, {"AFN", "؋"}, {"ALL", "L"}, ...]

```
```elixir
iex> CurrencyFormatter.get_currencies_for_select(:disambiguate_symbols)
[[{"AED", "د.إ"}, {"AFN", "؋"}, {"ALL", "Lek"}, ...]
```

Get the disambiguous symbol of a currrency
```elixir
iex> CurrencyFormatter.symbol(:AUD)
"A$"
```
## Installation

As this is [available in Hex](https://hex.pm/docs/publish), the package can be installed as:

1. Add currency_formatter to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [{:currency_formatter, "~> 0.4"}]
end
```

## Setup
By default you will have 172 currencies available, there are two configuration options that can be set in your app's `config.exs` file to limit the number of currencies.

If you would like to limit the list to only include specific currencies, you can configure a whitelist by passing a list of ISO codes to include:

```elixir
config :currency_formatter, :whitelist, ["EUR", "GBP", "USD"]
```

To exclude certain currencies from the list, you can configure a blacklist using a list of ISO codes to exclude:

```elixir
config :currency_formatter, :blacklist, ["XDR", "XAG", "XAU"]
```

## Documentation

API documentation is available at [https://hexdocs.pm/currency_formatter](https://hexdocs.pm/currency_formatter) and [http://smeevil.github.io/currency_formatter](http://smeevil.github.io/currency_formatter)




CHANGELOG.md
## 0.8.1 (2018-08-24)
- bumped deps

## 0.8.0 (2018-08-24)
- fixed formatting for negative cents, for example -5 and -53 would be displayed as €-,5 and €-,53. this has been remedied :)

## 0.7.0 (2018-02-22)
- you can now pass along the option `keep_decimals: true` with the formatters to ... wait for it... keep the decimals :)

example :
    assert "€1" == CurrencyFormatter.format(100, :eur)
    assert "€1,00" == CurrencyFormatter.format(100, :eur, keep_decimals: true)

- Added new methods to help with html formatting and warp the symbol and amount in spans with the class `currency-formatter-symbol` and `currency-formatter-amount`

    - `html_format/2`
    - `html_format/3`
    - `raw_html_format/2`
    - `raw_html_format/3`

## 0.6.0 (2018-02-14)
- Updated readme to cover the new disambiguate option
- Bumped deps

Thanks @duff for :
- For fixing a bug causing extraneous separator showing up at the front of negative values.
- Default behaviour is now to _not_ use disambiguate symbols, which can be turned on again by using an option like `CurrencyFormatter.format(123456, "AUD", disambiguate: true)`

## 0.5.0 (2017-10-06)

- Reformatted and cleaned upcode
- added dialyzer and more strict specs
- updated deps

## 0.4.8 (2017-05-29)
Thanks to @jknipp for adding blacklist support.
You can configure this in the settings just like you configure the whitelist

```elixir
config :currency_formatter, :blacklist, ["XDR", "XAG", "XAU"]
```

## 0.4.7 (2017-05-25)
Thanks to @jknipp for syncing up with the latest version of the currency_iso.json

## 0.4.6 (2017-04-04)

  - bumped deps

## 0.4.5 (2017-01-09)
Thanks to @marceldegraaf for removing deprecation warnings for elixir 1.4

## 0.4.4 (2016-12-07)
Added a configuration option so you can whitelist supported currencies. To make use of this you can update you app's config.exs with

```elixir
config :currency_formatter, :whitelist, ["EUR", "GBP", "USD"]
```

Thank you @vraravam for the suggestion :)

## 0.4.3 (2016-10-31)

  - Updated data source, returning usd by default. Thank you @optikfluffel !
  - bumped deps

## 0.4.2 (2016-08-11)

  - bumped deps

## 0.4.1 (2016-08-01)

  - Added a new function CurrencyFormatter.symbol/1 which will return the disambiguous currency symbol

## 0.4.0 (2016-06-25)

  - Added a new function (CurrencyFormatter.get_currencies_for_select) which returns a list with all currencies to be used in a select dropdown.

## 0.3.0 (2016-06-25)

  - Added a new function (CurrencyFormatter.get_currencies) which returns a map with all currencies and their formatting rules. Thank you @minibikini !
  - bumped all dependencies

## 0.2.0 (2016-06-18)
Added a new function (CurrencyFormatter.instructions(:EUR)) which returns formatting settings for a currency. You can use this information to create for example an input field to use with the Autonumeric js library.

## 0.0.1 (2016-03-23)

  - Initial commit


test/currency_formatter_test.exs
defmodule CurrencyFormatterTest do
  use ExUnit.Case, async: false

  doctest CurrencyFormatter

  setup do
    previous = Application.get_env(:currency_formatter, :whitelist)
    previous_blacklist = Application.get_env(:currency_formatter, :blacklist)

    on_exit(fn ->
      Application.put_env(:currency_formatter, :whitelist, previous)
      Application.put_env(:currency_formatter, :blacklist, previous_blacklist)
    end)

    :ok
  end

  test "should correctly format in euro and dollars" do
    assert "€0,01" == CurrencyFormatter.format(1, :eur)
    assert "$0.01" == CurrencyFormatter.format(1, "USD")
    assert "0,01Ft" == CurrencyFormatter.format(1, "huf")
    assert "$0.01" == CurrencyFormatter.format(1, :AUD)
  end

  test "it formats negative cents correctly" do
    assert "€-0,05" == CurrencyFormatter.format(-5, :eur)
    assert "$-0.05" == CurrencyFormatter.format(-5, "USD")
    assert "€-0,53" == CurrencyFormatter.format(-53, :eur)
    assert "$-0.53" == CurrencyFormatter.format(-53, "USD")
  end

  test "should format an amount_in_cents as integer to a price in dollars" do
    assert "$0.01" == CurrencyFormatter.format(1, :usd)
    assert "$0.12" == CurrencyFormatter.format(12, :usd)
    assert "$1.23" == CurrencyFormatter.format(123, :usd)
    assert "$12.34" == CurrencyFormatter.format(1234, :usd)
    assert "$123.45" == CurrencyFormatter.format(12345, :usd)
    assert "$1,234.56" == CurrencyFormatter.format(123_456, :usd)
    assert "$1,000" == CurrencyFormatter.format(100_000, :usd)
    assert "$-1,000" == CurrencyFormatter.format(-100_000, :usd)
    assert "$12,345.67" == CurrencyFormatter.format(1_234_567, :usd)
    assert "$-12,345.67" == CurrencyFormatter.format(-1_234_567, :usd)
    assert "$123,456.78" == CurrencyFormatter.format(12_345_678, :usd)
    assert "$1,234,567.89" == CurrencyFormatter.format(123_456_789, :usd)
    assert "$12,345,678.90" == CurrencyFormatter.format(1_234_567_890, :usd)
    assert "$-113,728" == CurrencyFormatter.format(-11_372_800, :usd)
    assert "$-499" == CurrencyFormatter.format(-49900, :usd)
    assert "$1" == CurrencyFormatter.format(100, :usd)
    assert "$1.00" == CurrencyFormatter.format(100, :usd, keep_decimals: true)
  end

  test "should format an amount_in_cents as integer to a price in euros" do
    assert "€0,01" == CurrencyFormatter.format(1, :eur)
    assert "€0,12" == CurrencyFormatter.format(12, :eur)
    assert "€1,23" == CurrencyFormatter.format(123, :eur)
    assert "€12,34" == CurrencyFormatter.format(1234, :eur)
    assert "€123,45" == CurrencyFormatter.format(12345, :eur)
    assert "€1.234,56" == CurrencyFormatter.format(123_456, :eur)
    assert "€1.000" == CurrencyFormatter.format(100_000, :eur)
    assert "€-1.000" == CurrencyFormatter.format(-100_000, :eur)
    assert "€12.345,67" == CurrencyFormatter.format(1_234_567, :eur)
    assert "€-12.345,67" == CurrencyFormatter.format(-1_234_567, :eur)
    assert "€123.456,78" == CurrencyFormatter.format(12_345_678, :eur)
    assert "€1.234.567,89" == CurrencyFormatter.format(123_456_789, :eur)
    assert "€12.345.678,90" == CurrencyFormatter.format(1_234_567_890, :eur)
    assert "€-113.728" == CurrencyFormatter.format(-11_372_800, :eur)
    assert "€-499" == CurrencyFormatter.format(-49900, :eur)
    assert "€1" == CurrencyFormatter.format(100, :eur)
    assert "€1,00" == CurrencyFormatter.format(100, :eur, keep_decimals: true)
  end

  test "should return html" do
    assert ~s[<span class="currency-formatter-symbol">€</span><span class="currency-formatter-amount">1.234</span>] ==
             CurrencyFormatter.raw_html_format(123_400, :eur)
    assert ~s[<span class="currency-formatter-symbol">€</span><span class="currency-formatter-amount">1.234,00</span>] ==
             CurrencyFormatter.raw_html_format(123_400, :eur, keep_decimals: true)
  end

  test "should format an amount_in_cents as string to a price" do
    assert "$0.01" == CurrencyFormatter.format("1", :usd)
    assert "$0.12" == CurrencyFormatter.format("12", :usd)
    assert "$1.23" == CurrencyFormatter.format("12.3", :usd)
    assert "$12.34" == CurrencyFormatter.format("12.34", :usd)
    assert "$123.45" == CurrencyFormatter.format("123,45", :usd)
    assert "$1,234.56" == CurrencyFormatter.format("1234,56", :usd)
    assert "$12,345.67" == CurrencyFormatter.format("12345.67", :usd)
    assert "$123,456.78" == CurrencyFormatter.format("123456,78", :usd)
    assert "$1,234,567.89" == CurrencyFormatter.format("1234567.89", :usd)
    assert "$12,345,678.90" == CurrencyFormatter.format("12345678,90", :usd)
    assert "$1" == CurrencyFormatter.format("1.00", :usd)
  end

  test "should use a currency symbol when given" do
    assert "$1,234,567.89" == CurrencyFormatter.format(123_456_789, :USD)
  end

  test "should allow the ability to disambiguate the currency symbol" do
    assert "US$12.34" == CurrencyFormatter.format(1234, :usd, disambiguate: true)
    assert "C$12.34" == CurrencyFormatter.format(1234, :cad, disambiguate: true)
    assert "C$12" == CurrencyFormatter.format(1200, :cad, disambiguate: true)
    assert "C$12.00" == CurrencyFormatter.format(1200, :cad, disambiguate: true, keep_decimals: true)

    assert "$12.34" == CurrencyFormatter.format(1234, :usd, disambiguate: false)
    assert "$12" == CurrencyFormatter.format(1200, :cad, disambiguate: false)
    assert "$12.00" == CurrencyFormatter.format(1200, :cad, disambiguate: false, keep_decimals: true)
  end

  test "should return a map with formatting instructions" do
    assert %{
             "alternate_symbols" => [],
             "decimal_mark" => ",",
             "html_entity" => "&#x20AC;",
             "iso_code" => "EUR",
             "iso_numeric" => "978",
             "name" => "Euro",
             "priority" => 2,
             "smallest_denomination" => 1,
             "subunit" => "Cent",
             "subunit_to_unit" => 100,
             "symbol" => "€",
             "symbol_first" => true,
             "thousands_separator" => "."
           } == CurrencyFormatter.instructions(:eur)

    assert %{
             "alternate_symbols" => [],
             "decimal_mark" => ",",
             "html_entity" => "&#x20AC;",
             "iso_code" => "EUR",
             "iso_numeric" => "978",
             "name" => "Euro",
             "priority" => 2,
             "smallest_denomination" => 1,
             "subunit" => "Cent",
             "subunit_to_unit" => 100,
             "symbol" => "€",
             "symbol_first" => true,
             "thousands_separator" => "."
           } == CurrencyFormatter.instructions("EUR")
  end

  test "should return formatting instructions for USD by default" do
    assert %{
             "smallest_denomination" => 1,
             "subunit" => "Cent",
             "subunit_to_unit" => 100,
             "symbol_first" => true,
             "alternate_symbols" => ["US$"],
             "decimal_mark" => ".",
             "html_entity" => "$",
             "iso_code" => "USD",
             "iso_numeric" => "840",
             "name" => "United States Dollar",
             "priority" => 1,
             "symbol" => "$",
             "thousands_separator" => ",",
             "disambiguate_symbol" => "US$"
           } == CurrencyFormatter.instructions()
  end

  test "should return the complete currencies list" do
    assert %{"usd" => %{}, "eur" => %{}} = CurrencyFormatter.get_currencies()
  end

  test "should return a list usable for select dropdowns" do
    assert ["AED", "AFN", "ALL"] = CurrencyFormatter.get_currencies_for_select() |> Enum.take(3)
  end

  test "should return a list usable for select dropdowns using names" do
    assert [
             {"AED", "United Arab Emirates Dirham"},
             {"AFN", "Afghan Afghani"},
             {"ALL", "Albanian Lek"}
           ] = CurrencyFormatter.get_currencies_for_select(:names) |> Enum.take(3)
  end

  test "symbol" do
    assert "$" == CurrencyFormatter.symbol(:AUD)
  end

  test "disambiguous_symbol" do
    assert "A$" == CurrencyFormatter.disambiguous_symbol(:AUD)
  end

  test "disambiguous_symbol default to symbol if the currency doesn't have a disambiguate_symbol" do
    assert "դր." == CurrencyFormatter.disambiguous_symbol(:amd)
  end

  test "should return a list usable for select dropdowns using currency symbols" do
    assert {"AUD", "$"} ==
             CurrencyFormatter.get_currencies_for_select(:symbols)
             |> Enum.find(fn {iso, _} -> iso == "AUD" end)
  end

  test "should return a list usable for select dropdowns using disambiguate currency symbols" do
    assert {"AUD", "A$"} ==
             CurrencyFormatter.get_currencies_for_select(:disambiguate_symbols)
             |> Enum.find(fn {iso, _} -> iso == "AUD" end)
  end

  test "should raise an error" do
    assert_raise RuntimeError,
                 ":other is not supported, please use either :names, :symbols or :disambiguate_symbols",
                 fn -> CurrencyFormatter.get_currencies_for_select(:other) end
  end

  test "get_currencies_for_select should only return whitelisted currencies" do
    Application.put_env(:currency_formatter, :whitelist, ["EUR", "USD", "CAD"])

    [{"CAD", "C$"}, {"EUR", "€"}, {"USD", "US$"}] =
      CurrencyFormatter.get_currencies_for_select(:disambiguate_symbols)
  end

  test "get_currencies should return all currencies except blacklisted currencies" do
    blacklist = ["XDR", "XAG", "XAU"]
    Application.put_env(:currency_formatter, :blacklist, blacklist)

    currencies = CurrencyFormatter.get_currencies()

    [] =
      currencies
      |> Enum.filter(fn {_, %{"iso_code" => iso_code}} -> Enum.member?(blacklist, iso_code) end)

    blacklist
    |> Enum.each(fn code ->
      assert Map.get(currencies, String.downcase(code)) == nil
    end)
  end
end


mix.exs
defmodule CurrencyFormatter.Mixfile do
  use Mix.Project

  def project do
    [
      app: :currency_formatter,
      version: "0.8.1",
      description: "A library to help with formatting a number to a currency using iso standards and other convenience functions related to formatting currencies",
      package: package(),
      elixir: "~> 1.7",
      build_embedded: Mix.env == :prod,
      start_permanent: Mix.env == :prod,
      deps: deps(),
      dialyzer: [
        plt_add_deps: true,
        ignore_warnings: "dialyzer.ignore-warnings",
        flags: [
          :error_handling,
          :race_conditions,
          :unknown,
          :unmatched_returns,
        ],
      ],
      preferred_cli_env: [
        coveralls: :test,
        "coveralls.detail": :test,
        "coveralls.post": :test,
        "coveralls.html": :test
      ],
      test_coverage: [
        tool: ExCoveralls
      ]
    ]
  end

  def application do
    [extra_applications: [:logger]]
  end

  defp deps do
    [
      {:dialyxir, ">= 0.0.0", only: :dev},
      {:earmark, ">= 0.0.0", only: :dev},
      {:ex_doc, ">= 0.0.0", only: :dev},
      {:excoveralls, ">= 0.0.0", only: :test},
      {:phoenix_html, ">= 0.0.0"},
      {:poison, "~> 3.1.0"},
    ]
  end
  defp package do
    [
      maintainers: ["Gerard de Brieder"],
      licenses: ["WTFPL"],
      files: ["lib", "mix.exs", "README*", "LICENSE*"],
      links: %{
        "GitHub" => "https://github.com/smeevil/currency_formatter",
        "Docs" => "http://smeevil.github.io/currency_formatter/"
      }
    ]
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure for your application as:
#
#     config :currency_formatter, key: :value
#
# And access this configuration in your application as:
#
#     Application.get_env(:currency_formatter, :key)
#
# Or configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"


lib/currency_formatter.ex
defmodule CurrencyFormatter do
  @moduledoc """
  This module takes care of formatting a number to a currency.
  You can also request a map containing all the formatting settings for a currency.
  """

  @currencies "./lib/currency_iso.json"
              |> File.read!()
              |> Poison.decode!()

  @doc """
  Formats a number to currency

  ## examples

      iex> CurrencyFormatter.format(123456, :usd)
      "$1,234.56"

      iex> CurrencyFormatter.format(654321, :eur)
      "€6.543,21"

      iex> CurrencyFormatter.format(654321, "AUD")
      "$6,543.21"

      iex> CurrencyFormatter.format(123456, "AUD", disambiguate: true)
      "A$1,234.56"

  """
  @spec format(String.t() | number | atom, String.t()) :: String.t()
  def format(number, currency, opts \\ [])

  def format(number, currency, opts) when is_atom(currency) do
    format(number, Atom.to_string(currency), opts)
  end

  def format(number, currency, opts) when is_integer(number) do
    number
    |> to_string
    |> format(currency, opts)
  end

  def format(number_string, currency, opts) when is_binary(number_string) and is_binary(currency) do
    format = instructions(currency)

    number_string
    |> remove_non_numbers
    |> add_subunit_separator
    |> add_padding
    |> split_units_and_subunits
    |> handle_cents(format, opts)
    |> set_symbol(format, opts)
  end

  @doc """
  Formats a number to currency wrapped in span tags with classes in a Phoenix.HTML safe way
  You can directly use this in your phoenix templates.

  ## example

      iex> CurrencyFormatter.html_format(123456, "EUR")
      [
        safe: [
          60,
          "span",
          [[32, "class", 61, 34, "currency-formatter-symbol", 34]],
          62,
          "€",
          60,
          47,
          "span",
          62
        ],
        safe: [
          60,
          "span",
          [[32, "class", 61, 34, "currency-formatter-amount", 34]],
          62,
          "1.234,56",
          60,
          47,
          "span",
          62
        ]
      ]
  """
  def html_format(number, currency, opts \\ Keyword.new()) do
    opts = Keyword.put_new(opts, :html, true)
    format(number, currency, opts)
  end

  @doc """
  Formats a number to currency wrapped in span tags with classes.
  This will return html as a string without any escaping. When using phoenix, consider using the `format_html/1` function

  ## example

      iex> CurrencyFormatter.raw_html_format(123456, "EUR")
      ~s[<span class="currency-formatter-symbol">€</span><span class="currency-formatter-amount">1.234,56</span>]

      iex> CurrencyFormatter.raw_html_format(123400, "EUR", keep_decimals: true)
      ~s[<span class="currency-formatter-symbol">€</span><span class="currency-formatter-amount">1.234,00</span>]
  """
  def raw_html_format(number, currency, opts \\ Keyword.new()) do
    number
    |> html_format(currency, opts)
    |> Enum.map(&Phoenix.HTML.safe_to_string/1)
    |> Enum.join("")
  end

  @doc """
  Returns a map with formatting settings for a currency

  ## examples

      iex> CurrencyFormatter.instructions(:EUR)
      %{"alternate_symbols" => [], "decimal_mark" => ",", "html_entity" => "&#x20AC;",
      "iso_code" => "EUR", "iso_numeric" => "978", "name" => "Euro", "priority" => 2,
      "smallest_denomination" => 1, "subunit" => "Cent", "subunit_to_unit" => 100,
      "symbol" => "€", "symbol_first" => true, "thousands_separator" => "."}

  """
  @spec instructions(atom) :: map
  def instructions(currency \\ :USD)

  def instructions(currency) when is_atom(currency) do
    currency
    |> Atom.to_string()
    |> instructions
  end

  @spec instructions(String.t()) :: map
  def instructions(currency) when is_binary(currency) do
    Map.get(@currencies, String.downcase(currency))
  end

  @doc """
  Returns a map containing the full list of currencies

  ## examples

      iex> currencies = CurrencyFormatter.get_currencies()
      iex> Enum.count(currencies)
      172
      iex> currencies["usd"]
      %{"alternate_symbols" => ["US$"], "decimal_mark" => ".",
      "disambiguate_symbol" => "US$", "html_entity" => "$", "iso_code" => "USD",
      "iso_numeric" => "840", "name" => "United States Dollar", "priority" => 1,
      "smallest_denomination" => 1, "subunit" => "Cent", "subunit_to_unit" => 100,
      "symbol" => "$", "symbol_first" => true, "thousands_separator" => ","}

  """
  @spec get_currencies() :: map
  def get_currencies do
    @currencies
    |> whitelist(Application.get_env(:currency_formatter, :whitelist))
    |> blacklist(Application.get_env(:currency_formatter, :blacklist))
  end

  @spec whitelist(map, nil | list) :: map

  defp whitelist(currencies, nil), do: currencies

  defp whitelist(currencies, whitelist) do
    Map.take(currencies, downcase(whitelist))
  end

  @spec blacklist(map, nil | list) :: map
  defp blacklist(currencies, nil), do: currencies

  defp blacklist(currencies, blacklist) do
    Map.drop(currencies, downcase(blacklist))
  end

  @spec downcase(list) :: list
  defp downcase(list) when is_list(list) do
    Enum.map(list, fn val -> String.downcase(val) end)
  end

  @doc """
  Returns a List with tuples that can be used as options for select

  ## Examples
      CurrencyFormatter.get_currencies_for_select()
      ["AED", "AFN", "ALL",...]

  """
  @spec get_currencies_for_select() :: [{String.t(), String.t()}]
  def get_currencies_for_select do
    get_currencies()
    |> Enum.map(fn {_, c} -> c["iso_code"] end)
    |> Enum.sort()
  end

  @doc """
  Returns a List with tuples that can be used as options for select

  ## Examples
      CurrencyFormatter.get_currencies_for_select(:names)
      [{"AED", "United Arab Emirates Dirham"}, {"AFN", "Afghan Afghani"} , {"ALL", "Albanian Lek"}, ...]

      CurrencyFormatter.get_currencies_for_select(:symbols)
      [{"AUD", "$"}, {"CAD", "$"}, {"USD", "$"},...]

      CurrencyFormatter.get_currencies_for_select(:disambiguate_symbols)
      [{"AUD", "A$"}, {"CAD", "C$"}, {"USD", "$"}, ...]
  """
  @spec get_currencies_for_select(atom) :: [{String.t(), String.t()}]
  def get_currencies_for_select(:names) do
    get_currencies()
    |> map_names
    |> Enum.sort()
  end

  def get_currencies_for_select(:symbols) do
    get_currencies()
    |> map_symbols
    |> Enum.sort()
  end

  def get_currencies_for_select(:disambiguate_symbols) do
    get_currencies()
    |> map_disambiguate_symbols
    |> Enum.sort()
  end

  def get_currencies_for_select(format) do
    raise(
      "#{inspect(format)} is not supported, please use either :names, :symbols or :disambiguate_symbols"
    )
  end

  @doc """
  Returns the symbol of a currency

  ## Example

      iex> CurrencyFormatter.symbol(:AUD)
      "$"

  """
  @spec symbol(atom) :: String.t()
  def symbol(currency) do
    currency
    |> CurrencyFormatter.instructions()
    |> get_symbol
  end

  @doc """
  Returns the disambiguous symbol of a currency

  ## Example

      iex> CurrencyFormatter.disambiguous_symbol(:AUD)
      "A$"

  """
  @spec symbol(atom) :: String.t()
  def disambiguous_symbol(currency) do
    currency
    |> CurrencyFormatter.instructions()
    |> get_disambiguous_symbol
  end

  @spec map_names(map) :: [map]
  defp map_names(map), do: Enum.map(map, fn {_, c} -> {c["iso_code"], c["name"]} end)

  @spec map_symbols(map) :: [map]
  defp map_symbols(map), do: Enum.map(map, fn {_, c} -> {c["iso_code"], c["symbol"]} end)

  @spec map_disambiguate_symbols(map) :: [map]
  defp map_disambiguate_symbols(map),
    do: Enum.map(map, fn {_, c} -> {c["iso_code"], c["disambiguate_symbol"] || c["symbol"]} end)

  @spec remove_non_numbers(String.t()) :: String.t()
  defp remove_non_numbers(string), do: String.replace(string, ~r/[^0-9-]/, "")

  @spec add_subunit_separator(String.t()) :: String.t()
  defp add_subunit_separator(string),
    do: String.replace(string, ~r/^0*([0-9-]+)(\d{2})$/, "\\1,\\2")

  @spec add_padding(String.t()) :: String.t()
  defp add_padding(""), do: "0,00"
  defp add_padding("-" <> centified) when byte_size(centified) == 1, do: "-0,0" <> centified
  defp add_padding(centified) when byte_size(centified) == 1, do: "0,0" <> centified
  defp add_padding("-," <> centified) when byte_size(centified) == 2, do: "-0," <> centified
  defp add_padding(centified) when byte_size(centified) == 2, do: "0," <> centified
  defp add_padding(centified), do: centified

  @spec split_units_and_subunits(binary) :: [binary]
  defp split_units_and_subunits(string), do: String.split(string, ",", parts: 2)

  @spec handle_cents(list, map, map) :: String.t()
  defp handle_cents([x, "00" = y], format, opts) do
    if Keyword.get(opts, :keep_decimals) do
      set_separators(x, format) <> format["decimal_mark"] <> y
    else
      set_separators(x, format)
    end
  end

  defp handle_cents([x, y], format, _opts),
    do: set_separators(x, format) <> format["decimal_mark"] <> y

  @spec set_separators(String.t(), map) :: String.t()
  defp set_separators(string, format) do
    string
    |> String.to_charlist()
    |> Enum.reverse()
    |> set_separators(format["thousands_separator"], "")
  end

  defp set_separators([a, b, c, ?- | tail], separator, acc) do
    set_separators([?- | tail], separator, [c, b, a | acc])
  end

  defp set_separators([a, b, c, d | tail], separator, acc) do
    set_separators([d | tail], separator, [separator, c, b, a | acc])
  end

  defp set_separators(list, _, acc) do
    list
    |> Enum.reverse()
    |> Kernel.++(acc)
    |> Kernel.to_string()
  end

  @spec set_symbol(String.t(), map, Keyword.t()) :: String.t()
  defp set_symbol(number_string, %{"symbol_first" => true} = config, opts) do
    symbol = get_symbol(config, opts)
    if Keyword.get(opts, :html) do
      wrap_in_spans(symbol: symbol, amount: number_string)
    else
      symbol <> number_string
    end
  end

  defp set_symbol(number_string, config, opts) do
    if Keyword.get(opts, :html) do
      spans = wrap_in_spans(amount: number_string, symbol: get_symbol(config, opts))
      Phoenix.HTML.safe_to_string(spans)
    else
      number_string <> get_symbol(config, opts)
    end
  end

  defp wrap_in_spans(symbol: symbol, amount: amount) do
    [wrap("currency-formatter-symbol", symbol), wrap("currency-formatter-amount", amount)]
  end

  defp wrap_in_spans(amount: amount, symbol: symbol) do
    [wrap("currency-formatter-amount", amount), wrap("currency-formatter-symbol", symbol)]
  end

  defp wrap(class, text) do
    Phoenix.HTML.Tag.content_tag(:span, text, class: class)
  end

  @spec get_symbol(map, keyword | nil) :: String.t()
  defp get_symbol(config, opts \\ Keyword.new()) do
    if Keyword.get(opts, :disambiguate),
      do: get_disambiguous_symbol(config),
      else: config["symbol"]
  end

  @spec get_disambiguous_symbol(map) :: String.t()
  defp get_disambiguous_symbol(%{"disambiguate_symbol" => symbol}), do: symbol
  defp get_disambiguous_symbol(config), do: config["symbol"]
end
